src/semant.ml

```diff
@@ -201,29 +201,50 @@ let rec check_expr expression table =
                    check_expr expr2 table, check_expr expr3 table))
     | TensorIdx(_,_) -> raise (Failure "Not yet implemented")
 
+(* If a function has a single type in its decl and the same
+ * id appears in its definition raise error, else return true*)
+let recursive_check (ftyp, fdef) =
+  let {types=ftyp'; _} = ftyp in
+  let single_typ = match ftyp' with
+    | Nat | Bool | Tensor(_) -> true
+    | Arrow(_) -> false in
+  let {ftyp_name=fid; _} = ftyp in
+  let rec rec_def fexpr = match fexpr with
+    | Literal(_) | Fliteral(_) | BoolLit(_) -> false
+    | TLit(e) -> List.for_all rec_def e
```

Comment:
TLits are always going to be lists of floats, which can never be recursively defined. The fact that they are represented as `expr list` in the Ast is simply a convenience for parsing reasons.

Reply:
Ok cool. Yeah, I hesitated when I wrote that for that reason. I'll bump it up to the line with the other lits.